### PR TITLE
feat/enterpriseportal: implement GetCodyGatewayUsage RPC

### PR DESCRIPTION
**Pull request by bobheadxi**
Uses the refactors from #63528 and the new RPC proposed in #63531 to implement the existing GraphQL query for getting Cody Gateway usage stats using Enterprise Portal. Also adds some basic instrumentation on `codygatewayevents.Service`.

Closes https://linear.app/sourcegraph/issue/CORE-201
Requires https://github.com/sourcegraph/managed-services/pull/1685

## Test plan

Without setting any new vars:

```sh
$ sg run enterprise-portal
# ...
[enterprise-...l] WARN service.codygatewayevents service/codygatewayevents.go:11 CodyGatewayEvents service is not configured
```

Add to `sg.config.overwrite.yaml`:

```yaml
  enterprise-portal:
    env:
      CODY_GATEWAY_EVENTS_PROJECT_ID: telligentsourcegraph
```

The client auths by default credentials, so if you have access to the dataset everything will "just work". Request S2 data by subscription ID at `:6081/debug/grpcui/`:

Screenshot 2024-06-28 at 1 55 49 PM

### CODE REVIEW

**Comment by bobheadxi** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63555?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#63555** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63555?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#63531** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63531?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bobheadxi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: cmd/enterprise-portal/service/config.go [hunk 1]
```diff
@@ -16,6 +18,9 @@ type Config struct {
 		IncludeProductionLicenses bool
 	}
 
+	// If nil, no connection was configured.
+	CodyGatewayEvents *codygatewayevents.ServiceBigQueryOptions
```
Comment (eseliger) [inline cmd/enterprise-portal/service/config.go:22]: Q: Should this be a panic case? 🤔

File: cmd/enterprise-portal/internal/codyaccessservice/v1.go [hunk 2]
```diff
@@ -143,3 +145,38 @@ func (s *handlerV1) ListCodyGatewayAccesses(ctx context.Context, req *connect.Re
 		log.Strings("accessedSubscriptions", accessedSubscriptions))
 	return connect.NewResponse(&resp), nil
 }
+
+func (s *handlerV1) GetCodyGatewayUsage(ctx context.Context, req *connect.Request[codyaccessv1.GetCodyGatewayUsageRequest]) (*connect.Response[codyaccessv1.GetCodyGatewayUsageResponse], error) {
+	logger := trace.Logger(ctx, s.logger)
+
+	// 🚨 SECURITY: Require approrpiate M2M scope.
+	requiredScope := samsm2m.EnterprisePortalScope("codyaccess", scopes.ActionRead)
```
Comment (eseliger) [inline cmd/enterprise-portal/internal/codyaccessservice/v1.go:153]: could we have a constant for "codyaccess" too? That way I could hover over it and see a docstring to understand what this does :D

File: cmd/enterprise-portal/internal/codyaccessservice/v1_store.go [hunk 3]
```diff
@@ -16,23 +24,106 @@ type StoreV1 interface {
...
+		embeddings, err := s.CodyGatewayEvents.EmbeddingsUsageForActor(
+			ctx,
+			codygatewayactor.ActorSourceEnterpriseSubscription,
+			subscriptionID,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "get embeddings usage")
+		}
+		return func(u *codyaccessv1.CodyGatewayUsage) {
+			u.EmbeddingsUsage = convertCodyGatewayUsageDatapoints(embeddings)
```
Comment (eseliger) [inline cmd/enterprise-portal/internal/codyaccessservice/v1_store.go:110]: cool pattern :)

**Review by eseliger — APPROVED** [review]

File: cmd/enterprise-portal/internal/codyaccessservice/v1.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/enterprise-portal/internal/codyaccessservice/v1.go:153]: Good call, I opened a PR upstream: https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/49

I'll circle back and update the existing callsites in a follow-up :)

**Comment by bobheadxi** [comment]
### Merge activity

* **Jul 2, 12:39 PM EDT**: @bobheadxi merged this pull request with [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/63555).

**Pull request merged**